### PR TITLE
Adds Result ADT from permafrost to Omnitool.

### PR DESCRIPTION
**Pull request by stephanh**
Added the Result concept from Permafrost to Omnitool since I want to do something similar to Hdfs for Hive in ebenezer.

### CODE REVIEW

**Comment by SamRoberts** [comment]
Can we change the error type to something sensible? Something with a decent monoid instance.

**Comment by stephanh** [comment]
Open to suggestions. I have kept it this way to maintain some backwards compatibility with how it is in Permafrost currently.

**Comment by SamRoberts** [comment]
Right, but isn't this a good opportunity to break backwards compatibility? Or not?

`List[String \/ Throwable]` is probably the simplest suggestion that comes to mind, and there is a straightforward translation from `String \&/ Throwable` to it.

You could add more structure (e.g., error hierarchies), but as of right now I don't have any evidence to suggest the complexity is worth it. But maybe hide the internal structure and just expose iteration capabilities as a method, so we could add in more power later if we needed it?

**Comment by stephanh** [comment]
How about the ability to append more error strings instead?

**Comment by SamRoberts** [comment]
Would prefer a more fundamental improvement, but that will be ok.

File: core/src/main/scala/au/com/cba/omnia/omnitool/Result.scala [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+      e      => Result.error(message, e),
+      (_, e) => Result.error(message, e)
+    )
+
+  /**
+    * Adds an additional error message. Useful for adding more context as the error goes up the stack.
+    * 
+    * The new message is prepended to any existing message.
+    */
+  def addMessage(message: String, seperator: String = " "): Result[A] =
```
Comment (SamRoberts) [inline core/src/main/scala/au/com/cba/omnia/omnitool/Result.scala]: seperator -> separator.

I would imagine that most of the time the default separator string should be `: `. Or at least `, `.

**Comment by SamRoberts** [comment]
Apart from minor comments, looks good. :+1:

File: core/src/main/scala/au/com/cba/omnia/omnitool/Result.scala (hunk 1 above)
Comment (stephanh) [inline core/src/main/scala/au/com/cba/omnia/omnitool/Result.scala]: Changed.

**Pull request merged**